File: src/plot/density_comparison.jl
using StatsBase
using CairoMakie

include("constants.jl")
include("../data/load_results.jl")


"""
    Expects all data to be saved in the same folder as the fasta file for each dataset. See `load_results` for details.
"""
function density_comparison(
    fasta_paths::Vector{String},
    data_names::Vector{String},
    data_colors::Vector{Symbol},
    target::String;
    direction::Symbol=:vertical, # :horizontal or :vertical
    save_dir::Union{String, Nothing}=nothing,
)
    ########## Get plot constans ##########
    load_list = LOAD[target]
    min_val = MIN_VAL[target]
    max_val = MAX_VAL[target]
    ticks = TICKS[target]
    threshold = THRESHOLD[target]
    offset = OFFSET[target]

    
    ########## Load data ##########
    all_dfs = map(fasta_path -> load_results(fasta_path; load=load_list), fasta_paths)
    all_data = map(df -> convert(Vector{Float64}, df[!, target]), all_dfs)


    ########## Density plot ##########

    if direction == :horizontal
        ########## Horizontal ##########
        fig = Figure()
        ax = Axis(
            fig[1, 1],
            title=target * " density",
            xticks=ticks,
            yticks=((1:length(data_names)) * offset, data_names),
        )

        for i in length(all_data):-1:1
            density!(
                ax,
                all_data[i],
                offset=offset * i,
                #boundary=(0.0, maximum(vcat(all_data...))),

                color=data_colors[i],
                #color=:x,
                #colormap=:thermal,
                #colorrange=(-5, 5),

                strokewidth=1,
                strokecolor=:black,
            )
        end
        xlims!(ax, low=min_val, high=max_val)

        isnothing(threshold) || vlines!(ax, threshold, color=:red, linestyle=:dash, linewidth=2)

        display(fig)

    elseif direction == :vertical
        ########## Vertical ##########
        fig = Figure()
        ax = Axis(
            fig[1, 1],
            title=target * " density",
            yticks=ticks,
            xticks=((1:length(data_names)) * offset, data_names),
        )

        for i in length(all_data):-1:1
            density!(
                ax,
                all_data[i],
                offset=offset * i,
                direction=:y,
                #boundary=(0.0, maximum(vcat(all_data...))),

                color=data_colors[i],
                #color=:x,
                #colormap=:thermal,
                #colorrange=(-5, 5),

                strokewidth=1,
                strokecolor=:black,
            )
        end
        ylims!(ax, low=min_val, high=max_val)

        isnothing(threshold) || hlines!(ax, threshold, color=:red, linestyle=:dash, linewidth=2)

        display(fig)
    end


    ########## Save plot ##########
    if !isnothing(save_dir)
        isdir(save_dir) || mkpath(save_dir)
        save(joinpath(save_dir, target * "_density_" * String(direction) * ".png"), fig)
    end 
end

File: src/plot/density_comparison.py
from __future__ import annotations

import os
import sys
from pathlib import Path
from typing import List, Optional, Sequence

import matplotlib

matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np
from scipy.stats import gaussian_kde

CURRENT_DIR = Path(__file__).resolve().parent
SRC_DIR = CURRENT_DIR.parent
if str(SRC_DIR) not in sys.path:
    sys.path.insert(0, str(SRC_DIR))

from data.load_results import load_results  # noqa: E402
from plot.constants import (  # noqa: E402
    LOAD,
    MAX_VAL,
    MIN_VAL,
    OFFSET,
    OFFSET_DEFAULT,
    THRESHOLD,
    TICKS,
)
from plot.target_config import auto_ticks, resolve_range  # noqa: E402


def _kde_curve(values: List[float], grid: np.ndarray) -> np.ndarray:
    arr = np.asarray(values, dtype=float)
    if arr.size == 0:
        return np.zeros_like(grid)
    if arr.size == 1 or np.allclose(arr, arr[0]):
        # gaussian_kde fails on zero variance; emit a narrow spike at the value.
        spike = np.zeros_like(grid)
        idx = int(np.argmin(np.abs(grid - arr[0])))
        spike[idx] = 1.0
        return spike
    kde = gaussian_kde(arr)
    return kde(grid)


def density_comparison(
    fasta_paths: Sequence[str],
    data_names: Sequence[str],
    data_colors: Sequence[str],
    target: str,
    *,
    direction: str = "vertical",
    dfs: Optional[Sequence] = None,
    save_dir: Optional[str] = None,
) -> None:
    """Ridge-style density plot of `target` across datasets.

    `direction='vertical'` lays each dataset's KDE along the y-axis, offset on
    x. `direction='horizontal'` does the inverse. Pass `dfs` (a per-dataset list
    of already-loaded DataFrames) to bypass fasta-based loading — used for
    structure targets.
    """
    if direction not in ("vertical", "horizontal"):
        raise ValueError(f"direction must be 'vertical' or 'horizontal', got {direction!r}")

    threshold = THRESHOLD.get(target)
    offset = OFFSET.get(target, OFFSET_DEFAULT)

    if dfs is None:
        dfs = [load_results(fp, load=LOAD[target]) for fp in fasta_paths]
    all_data: List[List[float]] = [
        [float(v) for v in df[target].dropna().tolist()] for df in dfs
    ]

    min_val, max_val = resolve_range(target, MIN_VAL, MAX_VAL, all_data)
    ticks = TICKS.get(target)
    if ticks is None:
        ticks = auto_ticks(min_val, max_val)

    grid = np.linspace(min_val, max_val, 512)

    fig, ax = plt.subplots()
    n = len(all_data)
    label_positions = [(i + 1) * offset for i in range(n)]

    for i in range(n - 1, -1, -1):
        density = _kde_curve(all_data[i], grid)
        baseline = (i + 1) * offset
        if direction == "vertical":
            ax.fill_betweenx(
                grid,
                baseline,
                baseline + density,
                color=data_colors[i],
                edgecolor="black",
                linewidth=1,
            )
        else:
            ax.fill_between(
                grid,
                baseline,
                baseline + density,
                color=data_colors[i],
                edgecolor="black",
                linewidth=1,
            )

    ax.set_title(target + " density")

    if direction == "vertical":
        ax.set_yticks(ticks)
        ax.set_xticks(label_positions)
        ax.set_xticklabels(data_names)
        ax.set_ylim(min_val, max_val)
        ax.yaxis.grid(True, linestyle="--", color="gray")
        ax.set_axisbelow(True)
        if threshold is not None:
            ax.axhline(threshold, color="red", linestyle="--", linewidth=2)
    else:
        ax.set_xticks(ticks)
        ax.set_yticks(label_positions)
        ax.set_yticklabels(data_names)
        ax.set_xlim(min_val, max_val)
        ax.xaxis.grid(True, linestyle="--", color="gray")
        ax.set_axisbelow(True)
        if threshold is not None:
            ax.axvline(threshold, color="red", linestyle="--", linewidth=2)

    fig.tight_layout()

    if save_dir is not None:
        os.makedirs(save_dir, exist_ok=True)
        fig.savefig(os.path.join(save_dir, target + "_density_" + direction + ".png"))

    plt.close(fig)
